File: java/bigInt.java
import java.util.Scanner;
import java.math.BigInteger;

// http://www.codechef.com/NOV13/problems/SPOTWO/
public class bigInt  {
	static final int MOD = 1000000007;
	public static void main(String[] args) {
		doIt();
	}

	static void doIt() {
		Scanner sc = new Scanner(System.in);
		// PrintWriter pw = new PrintWriter(System.out);
		int t, n = 0;
		String bin = "";
		long ans = 0L;
		t = sc.nextInt();
		for(int ii = 0; ii < t; ii++){
			n = sc.nextInt();
			bin = Integer.toString(n, 2);
			ans = pow(2, new BigInteger(bin), MOD);
			ans = ans * ans % MOD;
			System.out.println(ans);
		}

	}

	static long pow(int n1, BigInteger p, int m){
		long n = n1;
		long ans = 1L;
		while (!p.equals(BigInteger.ZERO)) {
			// if((p&1) == 1) ans = (ans*n) % m;
			if(p.getLowestSetBit() == 0) ans = (ans*n) % m;
			n = (n*n) % m;
			// p = p>>1;
			p = p.shiftRight(1);
		}
		return ans;
	}

}

File: scala/bigint.scala
import java.util.Scanner;
import java.math.BigInteger;

object Solution  {
  val MOD = 1000000007;
  def main(args: Array[String]){
    doIt()
  }

  def doIt(){
    val sc = new Scanner(System.in);
    val t = sc.nextInt();
    0 until t foreach{ ii =>
      val as = sc.next();
      val bs = sc.next();
      val a = new BigInteger(as).mod(new BigInteger("1000000007")).longValue();
      var b = new BigInteger(bs).mod(new BigInteger("1000000006")).longValue();
      println(pow(a.toInt, b.toInt, MOD));
    }
  }
  def pow(n1:Int, p1:Int, m:Int):Int = {
    var n = n1.toLong;
    var p = p1;
	var ans = 1L;
	while (p != 0) {
	  if((p&1) == 1) ans = (ans*n) % m;
	  n = (n*n) % m;
	  p = p>>1;
	}
	return ans.toInt;
  }
}
